packages/storybook-builder-vite/vite-config.js

```diff
@@ -47,7 +47,10 @@ module.exports.pluginConfig = function pluginConfig(options, type) {
 
     if (type === 'development') {
         if (framework === 'react') {
-            plugins.push(require('@vitejs/plugin-react-refresh')());
+            plugins.push(require('@vitejs/plugin-react-refresh')({
+                // Do not treat story files as HMR boundaries, storybook itself needs to handle them.
+                exclude: [/\.stories\.(t|j)sx?$/, /node_modules/],
```

Comment:
Probably, also excluding `node_modules` would lead to bugs with linked dependencies which HMR should be managed by vite

Reply:
There should be also ability to override the exclude, as not everyone names files as `stories`

Reply:
>Probably, also excluding node_modules would lead to bugs with linked dependencies which HMR should be managed by vite

Excluding node_modules is the default behavior of plugin-react-refresh, so I wanted to keep that behavior for now, until/unless we hear of problems.

>There should be also ability to override the exclude, as not everyone names files as stories

There is, it is possible to change any of this config through the use of `viteFinal` in the storybook configuration file.

Reply:
It's also possible to read the list of stories (or the list of file globs) from the `config` variable, I think?

Reply:
Unfortunately it looks like `stories` from the config is not exposed to the builder. I logged out the `options` and didn't see it, and it's not in the type definitions, from what I can tell: https://github.com/storybookjs/storybook/blob/next/lib/core-common/src/types.ts#L164

Reply:
@shilman should that config option maybe be exposed in the builder API?

Reply:
I believe you can get it from options.presets.apply('stories') but I didn't test

Reply:
After looking into this, it's true that we can get the `stories` as @shilman suggested.  However, those are micromatch globs, and @vitejs/react-fast-refresh uses picomatch, which does not allow brace expansion.  We would need to resolve the entire list of files and pass those to the plugin.  Since this whole approach is a bit of a bandaid, I propose that we hold off on any more complicated exclusion logic until/unless we hear of problems.  Maybe we'll find a better solution entirely that works for more than just the react framework.